Lab №3 initial commit

xd6d opened the pull request.

xd6d committed:
Lab №3 initial commit

nosovk commented on Lab3-Node.js-Basics/package.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "lab3-node.js-basics",
+  "version": "1.0.0",
```
"type": "module" forgotten

nosovk commented on Lab3-Node.js-Basics/package.json:
```diff
@@ -0,0 +1,21 @@
+    "typescript": "^5.0.4"
+  },
+  "dependencies": {
+    "axios": "^1.4.0",
```
this packet not used at backend, you can use undici (it's included in node 20) or undici.fetch if you need fetch interface

nosovk commented on Lab3-Node.js-Basics/src/Task1.ts:
```diff
@@ -0,0 +1,20 @@
+
+async function runSequent<T>(array: Array<T>, fn: (item: T, index: number) => Promise<Awaited<{ item: T; index: number }>>) {
+    const res = []
+    for (let i = 0; i<array.length; i++){
```
for await?

nosovk replied:
for await это отдельная синтаксическая конструкция, с ней будет короче

nosovk commented on Lab3-Node.js-Basics/src/Task2.ts:
```diff
@@ -0,0 +1,15 @@
+
+function arrayChangeDelete<T>(array: T[], toDelete: (item: T) => boolean) {
+    let result = array.filter(toDelete)
+    let index = array.findIndex(toDelete)
+    while (index != -1) {
+        array.splice(index, 1)
+        index = array.findIndex(toDelete)
+    }
```
возможно как-то проще можно удалить элементы в массиве? Или хотя бы сократить этот код?

nosovk commented on Lab3-Node.js-Basics/src/Task3.ts:
```diff
@@ -0,0 +1,14 @@
+import axios from 'axios';
+
+const fs = require('fs')
```
node:fs, node:path etc...

nosovk replied:
для fs кстати есть промисифицированная версия

nosovk commented on Lab3-Node.js-Basics/src/Task3.ts:
```diff
@@ -0,0 +1,14 @@
+const path = require('path')
+
+
+const input = process.argv[2]
```
не хватает проверок на то что этот параметр есть

nosovk commented on Lab3-Node.js-Basics/src/Task3.ts:
```diff
@@ -0,0 +1,14 @@
+const input = process.argv[2]
+const filename = path.parse(input).name
+const dirPath = path.join(__dirname, `${filename}_pages`)
+fs.mkdir(dirPath, () => {})
```
в целом не хватает проверки ошибок

nosovk commented on Lab3-Node.js-Basics/tsconfig.json:
```diff
@@ -0,0 +1,109 @@
+
+    /* Language and Environment */
+    "target": "es2016",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
+    "lib": ["es6","dom"],                                /* Specify a set of bundled library declaration files that describe the target runtime environment. */
```
у нас же нет тут dom...

nosovk commented on Lab3-Node.js-Basics/tsconfig.json:
```diff
@@ -0,0 +1,109 @@
+    // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
+
+    /* Language and Environment */
+    "target": "es2016",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
```
можно что-то поближе к 2023 взять

nosovk commented on Lab3-Node.js-Basics/tsconfig.json:
```diff
@@ -0,0 +1,109 @@
+    // "moduleDetection": "auto",                        /* Control what method is used to detect module-format JS files. */
+
+    /* Modules */
+    "module": "commonjs",                                /* Specify what module code is generated. */
```
use esm pls

xd6d committed:
code simplification

xd6d committed:
added "type": "module"
removed axios dependency
added undici dependency
refreshed tsconfig.json

xd6d committed:
Task1.ts refactor

xd6d committed:
Task3.ts
require() changed to imports
added argument existence check
migrated from axios to undici

xd6d commented:
Перепрошую за бардак. Не знав, що на один бранч лише один пул реквест і всі пуши будуть падати сюди...💀💀
4 лабораторна робота вже у іншому бранчі.

nosovk commented on Lab3-Node.js-Basics/src/Task2.ts:
```diff
@@ -0,0 +1,12 @@
+
+function arrayChangeDelete<T>(array: T[], toDelete: (item: T) => boolean) {
+    let result = array.filter(toDelete)
+    while (array.findIndex(toDelete) != -1)
```
мы тут дважды ищем одно и то же, лучше так не делать, поиск дорогая операция.

nosovk replied:
вообще кривая реализация. Можно за один проход решить судьбу каждого элемента исходного масива

nosovk commented on Lab3-Node.js-Basics/src/Task3.ts:
```diff
@@ -0,0 +1,16 @@
+const filename = path.parse(input).name
+const dirPath = path.join(process.cwd(), `${filename}_pages`)
+fs.mkdir(dirPath, () => {})
+const links = JSON.parse(fs.readFileSync(input).toString())
```
лучше не использовать Sync метода. Есть node:fs/promise - промифицированная версия прямо стандартно есть. И ее можно через promise.all запускать в параллельно выполнение пачку операций сразу

nosovk commented on Lab3-Node.js-Basics/src/Task3.ts:
```diff
@@ -0,0 +1,16 @@
+const dirPath = path.join(process.cwd(), `${filename}_pages`)
+fs.mkdir(dirPath, () => {})
+const links = JSON.parse(fs.readFileSync(input).toString())
+for (let i = 0; i<links.length; i++) {
```
вот тут лучше переписать, на map и аналоги, чтобы запускать асинхронные задачи не дожидаясь завершения предыдущих

nosovk reviewed (changes requested):
fix comments

xd6d committed:
fixes
